src/main/java/org/elasticsearch/plugins/repository/ChainedPluginRepository.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins.repository;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Collection;
+
+public class ChainedPluginRepository implements PluginRepository {
+
+    private final String name;
+
+    private final ImmutableMap<String, PluginRepository> repositories;
+
+    private ChainedPluginRepository(String name, ImmutableMap<String, PluginRepository> repositories) {
+        this.name = name;
+        this.repositories = repositories;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public Collection<PluginDescriptor> find(String name) {
+        ImmutableList.Builder<PluginDescriptor> plugins = ImmutableList.builder();
+
+        if (repositories != null) {
```

Comment:
I think this could not be `null`. You can only build a `ChainedPluginRepository` using the builder which creates an `ImmutableMap`.